### PR TITLE
Add a middleware for comparator to use

### PR DESCRIPTION
**Pull request by bowenxia**
**What changed?**
Add a middleware for comparator to use.

**Why?**
We need to receive the yarpc req sent from PinotResponseComparator, and then pass the cadence-visibility-override value into ctx. So that PinotVisibilityTripleManager can use it to switch between pinot and es. 

**How did you test it?**
Will do manually testing. 

**Potential risks**

**Release notes**

**Documentation Changes**

### CODE REVIEW

File: common/rpc/middleware.go [hunk 1]
```diff
@@ -115,6 +116,28 @@ func (m *InboundMetricsMiddleware) Handle(ctx context.Context, req *transport.Re
...
+	if yarpcKey == persistence.Primary {
+		ctx = ContextWithOverride(ctx, persistence.Primary)
+	} else if yarpcKey == persistence.Secondary {
+		ctx = ContextWithOverride(ctx, persistence.Secondary)
+	}
+	return h.Handle(ctx, req, resw)
+}
+
+// ContextWithOverride adds a value in ctx
+func ContextWithOverride(ctx context.Context, value string) context.Context {
```
Comment (davidporter-id-au) [inline common/rpc/middleware.go:137]: this just needs to use a name which indicates it's specifically for the Visibility override. Consider a name like `contextWithVisibilityOverride`

File: common/rpc/middleware.go [hunk 2]
```diff
@@ -115,6 +116,28 @@ func (m *InboundMetricsMiddleware) Handle(ctx context.Context, req *transport.Re
...
+// YarpcKey is the const for yarpc key
+const ComparatorYarpcKey = "cadence-visibility-override"
+
+// PinotComparatorMiddleware checks the header of a grpc request, and then override the context accordingly
+// note: for Pinot Migration only (Jan. 2024)
+type PinotComparatorMiddleware struct{}
+
+func (m *PinotComparatorMiddleware) Handle(ctx context.Context, req *transport.Request, resw transport.ResponseWriter, h transport.UnaryHandler) error {
+	yarpcKey, _ := req.Headers.Get(ComparatorYarpcKey)
+	if yarpcKey == persistence.Primary {
```
Comment (davidporter-id-au) [inline common/rpc/middleware.go:128]: do you mind changing this to be something indicating it's intent to visibility? `Primary` and `Secondary` are a bit general, perhaps `VisibilityOverridePrimary` ...

**Review by davidporter-id-au — APPROVED** [review]

**Comment by bowenxia** [comment]
Note: will not merge it until I test it is working

**Comment by davidporter-id-au** [comment]
oh, be sure to run the linter too

**Review by davidporter-id-au — APPROVED** [review]

**Pull request merged**